docs/source/howto/codes.rst

```diff
@@ -3,173 +3,199 @@
 
  #. :py:meth:`~aiida.engine.processes.calcjobs.calcjob.CalcJob.define`
  #. :py:meth:`~aiida.engine.processes.calcjobs.calcjob.CalcJob.prepare_for_submission`
 
-We will now show how each of these can be implemented.
+Defining the spec
+-----------------
 
-Defining the specifications
----------------------------
-
-The |define| method is where one specifies the different inputs that the caller of the |CalcJob| will have to provide in order to run the code, as well as the outputs that will be produced (exit codes will be :ref:`discussed later<how-to:codes:parsing:errors>`).
-This is done through an instance of :py:class:`~aiida.engine.processes.process_spec.CalcJobProcessSpec`, which, as can be seen in the snippet below, is passed as the |spec| argument to the |define| method.
-For the code that adds up two numbers, we will need to define those numbers as inputs (let's call them ``x`` and ``y`` to label them) and the result as an output (``sum``).
-The snippet below shows one potential implementation, as it is included in ``aiida-core``:
+The |define| method tells AiiDA which inputs the |CalcJob| expects and which outputs it produces (exit codes will be :ref:`discussed later<how-to:codes:parsing:errors>`).
+This is done through an instance of the :py:class:`~aiida.engine.processes.process_spec.CalcJobProcessSpec` class, which is passed as the |spec| argument to the |define| method.
+For example:
 
 .. literalinclude:: ../../../aiida/calculations/arithmetic/add.py
     :language: python
     :pyobject: ArithmeticAddCalculation.define
-    :dedent: 4
 
-The first line of the |define| implementation calls the method of the parent class |CalcJob|.
-This step is crucial as it will define `inputs` and `outputs` that are common to all |CalcJob|'s and failing to do so will leave the implementation broken.
-After the super call, we modify the default values for some of these inputs that are defined by the base class.
-Inputs that have already been defined can be accessed from the |spec| through the :py:attr:`~plumpy.process_spec.ProcessSpec.inputs` attribute, which behaves like a normal dictionary.
+The first line of the method calls the |define| method of the |CalcJob| parent class.
+This necessary step defines the `inputs` and `outputs` that are common to all |CalcJob|'s.
+
+Next, we use the :py:meth:`~plumpy.process_spec.ProcessSpec.input` method in order to define our two input numbers ``x`` and ``y`` (we support integers and floating point numbers), and we use :py:meth:`~plumpy.process_spec.ProcessSpec.output` to define an output of the calculation with label ``'sum'``.
+Once a calculation finishes, any output node specified here will be linked to the calculation node with the specified link label.
+
+.. tip::
+
+    In real-world plugins, popular types of input nodes are Python dictionaries (:py:class:`~aiida.orm.nodes.data.dict.Dict`) and files (:py:class:`~aiida.orm.nodes.data.singlefile.SinglefileData`).
```

Comment:
I'm not quite sure what this tip is trying to tell me. Should I prefer those types?

Reply:
People go to this howto with the purpose of wrapping their own code, for which they will likely need Dict or SinglefileData input nodes.
This is just to give them an idea, happy to reformulate to make this more clear

Reply:
I think a good formulation would be along the lines of "Consider to use x y, and z as input parameter types, that are used in many real world ..."